### ISSUE TITLE
No global RBG construction in PSA

### ISSUE THREAD

**Issue by jethrogb** [Mbed-TLS/mbedtls#7685]
As discussed on [the mailing list](https://lists.trustedfirmware.org/archives/list/mbed-tls@lists.trustedfirmware.org/message/RJ7YPNBNWUNW2ICQJ72H2JMKPDKGQOLT/), MbedTLS plans to move from a fully dependency-injection based API (well, besides the memory allocator) to an API that using a global RBG construction. This makes it exceedingly difficult to use multiple RBG constructions within the same application. Consider the following uses:

#### Using multiple different RBG constructions

RBG constructions may be used for different purposes. For example:

* key generation
* cryptographic operations (e.g. IV generation, signing)
* runtime protection (blinding)

Different purposes can have different compliance requirements, and different compliance requirements can result in different performance-security trade-offs. As a concrete example, it may be acceptable to use a fast hardware instruction for blinding, but key generation for long-lived keys could require a slower more complicated construction.

I realize the current MbedTLS API doesn't fully distinguish between the different purposes, but you can at least pass in a construction that meets the most stringent requirements for the operation that's currently being done.

Related is the case where an application builder is integrating multiple different components, each of which relies on MbedTLS but has different expectations about the RBG construction.

#### Doing specific operations with specific RBG constructions & seeds

Even when dealing with randomness, there are many situations where you want to use a specific RBG construction and seed to deterministically arrive at the same results multiple times. For example:

* multiple actors in a distributed system agreeing on the same secret material
* a single actor with limited storage for long-lived keys, storing seeds instead of full keys (e.g. TPM CreatePrimary)
* known-answer tests

---

While mutable global or thread-local state and linking tricks may allow you to do these things, I consider these really just workarounds for something that is already working entirely correctly today. Especially when also considering integrations with other languages and build systems, this is just not "playing nice" with the ecosystem. For example, the Rust language and build system have taken a very strong stance against global state in libraries, and as such have very poor support for any global state and global configuration.

---

Fortanix's interest in MbedTLS is as follows:
* We maintain the [MbedTLS bindings](https://crates.io/crates/mbedtls) for the Rust language
* We implement a secure element using the cryptography code from the MbedTLS library (you could imagine we'd write a PSA opaque driver)
* We use the TLS implementation as our primary TLS implementation

**Comment by gilles-peskine-arm** [comment]
> multiple actors in a distributed system agreeing on the same secret material
>
> a single actor with limited storage for long-lived keys, storing seeds instead of full keys (e.g. TPM CreatePrimary)

The PSA API has a key derivation API for that. In Mbed TLS, it's implemented for every key type except RSA, for which the legacy API doesn't have a reproducible derivation. (`mbedtls_rsa_gen_key` is only reproducible for a given build of Mbed TLS; a security update might consume the RNG argument differently, for example if we add more blinding to plug a timing leak.)

**Comment by gilles-peskine-arm** [comment]
Regarding using an RNG for different purposes: even if we extend the PSA API to pass an explicit context that includes the RNG configuration and state, the same RNG would still be passed to underlying function. More work would be needed to support different configurations.

**Comment by gilles-peskine-arm** [comment]
Here's a very high-level overview of how I envisage an API extension to pass a crypto context around instead of maintaining a global state. The new API would coexist with the existing API.

* Functions that work on an existing operation keep the same prototypes. The operation object contains a reference to the crypto context.
* In extended builds, `mbedtls_svc_key_id_t` (which is `psa_key_id_t` when building for the current PSA API) is now a struct `{context, id_within_context}`. This means you don't have the standard API when the extended API is present: you have a modified API where you can't pass an integer with a key id (this is already the case when `MBEDTLS_PSA_CRYPTO_KEY_ID_ENCODES_OWNER` is enabled — that option is meant to be used to implement a crypto service that exposes the PSA crypto API via RPC).
* Key creation functions take an extra argument to convey the context. Unless we pass the context through the key attributes?
* `psa_crypto_init` and `mbedtls_psa_crypto_free` take extra argument to create/destroy a context. There may be extra arguments or functions to configure a context before/when it's created.
* No mixing of contexts is possible. E.g. `psa_key_derivation_output_key` can't create a key in a different context from where the operation was started.
* Driver support is still decided at compile time.
* The build option would also apply to the TLS stack, and would add another function to associate a context with an `mbedtls_ssl_config`.

Note that this doesn't do everything you're asking for: a given context would still have a single RNG. We can build on it to allow configuring a context with multiple RNG (though I don't really see the point).

**Comment by jethrogb** [comment]
@gilles-peskine-arm is your current plan (outside of what's been discussed in this issue) that things like the Cipher and Pk interfaces will also have a direct dependency on PSA?

**Comment by gilles-peskine-arm** [comment]
@jethrogb Yes, in Mbed TLS 4.0, there will no longer be a pk interface that's independent from PSA. For cipher there may be no `cipher.h` at all, or a compatibility layer that calls PSA functions.
